Update `no_mangle` to use the attribute template

ehuss opened the pull request:
New rules:
- ❗ `abi.no_mangle.syntax`
- ❗ `abi.no_mangle.allowed-positions`
- ❗ `abi.no_mangle.duplicates`
- ❗ `abi.no_mangle.export_name`
- ❗ `abi.no_mangle.ascii-only`
- ❗ `abi.no_mangle.generic`

ehuss committed:
Move abi.no_mangle.publicly-exported and tweak wording

This moves it below to match the template of having behavioral rules
after the syntax.

ehuss committed:
Move abi.no_mangle.unsafe and tweak wording

This is to match the template order.

ehuss committed:
Update the no_mangle introduction

It doesn't really work with any item, it is functions and statics.

ehuss committed:
Add no_mangle template rules

ehuss committed:
Add abi.no_mangle.export_name

This specifies the conflict between no_mangle and export_name.

ehuss committed:
Add abi.no_mangle.ascii-only

This is a restriction that wasn't previously documented.

ehuss committed:
Move no_mangle example into block

ehuss committed:
Add abi.no_mangle.generic

traviscross committed:
Revise `no_mangle` attribute text

Let's revise this section editorially.

RalfJung commented on src/abi.md:
```diff
@@ -61,27 +61,66 @@ Only the first use of `used` on an item has effect.
+`no_mangle` may only be used on items with a name that only contains ASCII characters.
+
+r[abi.no_mangle.generic]
+`no_mangle` has no effect on generic items.
```
With https://github.com/rust-lang/rust/pull/154585, no_mangle is forbidden on generic items.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/1904/3fde038352015967da1e417115ff511c8f71baf5..bcd64d89087c7619c8f51d84e6d5868dc65337ab)*

ehuss committed:
Update no_mangle when used with generic items

This restriction is being added in
https://github.com/rust-lang/rust/pull/154585.
